File: Kalorie/Kalorie/Core/UseCases/FetchMyCreatedMealsUseCase.swift
//
//  FetchMyCreatedMealsUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 20.08.2026.
//

import Foundation

protocol FetchMyCreatedMealsUseCaseProtocol {
    func callAsFunction() async throws -> [MyCreatedMealDomain]
}

struct FetchMyCreatedMealsUseCase: FetchMyCreatedMealsUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction() async throws -> [MyCreatedMealDomain] {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        let dtos: [MyCreatedMealDTO] = try await dataProvider.loadAsync(
            from: Constants.Firestore.myCreatedMeals(userId: userId),
            orderBy: "updated_at",
            descending: true,
            limit: 50
        )
        return dtos.map { $0.asDomain() }
    }
}

#if DEBUG
struct FetchMyCreatedMealsUseCaseFake: FetchMyCreatedMealsUseCaseProtocol {

    // MARK: - Properties

    var stubbedMeals: [MyCreatedMealDomain] = []
    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction() async throws -> [MyCreatedMealDomain] {
        if shouldThrow { throw URLError(.unknown) }
        return stubbedMeals
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/FetchMyCreatedMealsUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.MyCreatedMealDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.MyCreatedMealDTO
import antoni.kalorie.core.networking.loadAsync
import antoni.kalorie.core.utils.Constants

interface FetchMyCreatedMealsUseCaseProtocol {
    suspend operator fun invoke(): List<MyCreatedMealDomain>
}

class FetchMyCreatedMealsUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : FetchMyCreatedMealsUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(): List<MyCreatedMealDomain> {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        val dtos: List<MyCreatedMealDTO> = dataProvider.loadAsync(
            from = Constants.Firestore.myCreatedMeals(userId),
            orderBy = "updated_at",
            descending = true,
            limit = 50,
        )
        return dtos.map { it.asDomain() }
    }
}
